server/plugin.go

```diff
@@ -635,22 +706,87 @@ func (p *Plugin) HasGroupHook(ctx context.Context, user *gitlab.UserInfo, namesp
 	return found, err
 }
 
-func (p *Plugin) checkAndRefreshToken(token *oauth2.Token) (*oauth2.Token, error) {
-	// If there is only one minute left for the token to expire, we are refreshing the token.
-	// The detailed reason for this can be found here: https://github.com/golang/oauth2/issues/84#issuecomment-831492464
-	// We don't want the token to expire between the time when we decide that the old token is valid
-	// and the time at which we create the request. We are handling that by not letting the token expire.
-	if time.Until(token.Expiry) <= 1*time.Minute {
-		conf := p.getOAuthConfig()
-		src := conf.TokenSource(context.Background(), token)
-		newToken, err := src.Token() // this actually goes and renews the tokens
-		if err != nil {
-			return nil, errors.Wrap(err, "unable to get the new refreshed token")
+func (p *Plugin) refreshToken(userInfo *gitlab.UserInfo, token *oauth2.Token) (*oauth2.Token, error) {
+	conf := p.getOAuthConfig()
+	src := conf.TokenSource(context.Background(), token)
+
+	newToken, err := src.Token() // this actually goes and renews the tokens
+
+	if err != nil {
+		if strings.Contains(err.Error(), "\"error\":\"invalid_grant\"") {
+			p.handleRevokedToken(userInfo)
 		}
-		if newToken.AccessToken != token.AccessToken {
-			return newToken, nil
+		return nil, errors.Wrap(err, "unable to get the new refreshed token")
+	}
+
+	if newToken.AccessToken != token.AccessToken {
+		p.API.LogDebug("Gitlab token refreshed.", "UserID", userInfo.UserID)
```

Comment:
This may get logged too often. I'm thinking we should remove this. @hanzei What do you think?

Reply:
I plan to add a trace logger that only log to debug when a specific setting is activated. This would be a good log message for the trace logger.

Let's add a comment to do that later.